File: 백준/Gold/20040. 사이클 게임/사이클 게임.cc
// 21 23
#include<bits/stdc++.h>
using namespace std;

int n, m;
int total = 0;
vector<int> v(1000005, -1);
bool check = false;

int fin(int c){
    if(v[c]<0) return c;
    return v[c] = fin(v[c]);
}

bool uni(int a, int b, int num){
    int fir = fin(a);
    int sec = fin(b);
    if(fir == sec){
        cout<<num+1;
        return 1;
    }
    v[sec] = fir;
    return 0;
}

int main(void){
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin >> n >> m;
    
    for(int i=0; i<m; i++){
        int a, b;
        cin >>a >> b;
        if(uni(a, b, i)) return 0;

    }
    cout<< 0;
}

File: 백준/Gold/20040. 사이클 게임/사이클 게임.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int[] arr;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        arr = new int[n];
        Arrays.fill(arr, -1);
        boolean chk = false;
        for(int i=0; i<m; i++){
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            if(uni(a,b,i)){
                chk = true;
                break;
            }
        }
        if(!chk) System.out.println(0);
    }
    public static boolean uni(int a, int b, int num){
        int fir = fin(a);
        int sec = fin(b);
        if(fir == sec){
            System.out.println(num+1);
            return true;
        }
        arr[sec] = fir;
        return false;
    }
    public static int fin(int c){
        if(arr[c] < 0) return c;
        return arr[c] = fin(arr[c]);
    }
}
